### PR TITLE
Add ES256 and PS256 signing algorithm support

### PR DESCRIPTION
**Pull request by zshenker**
## Summary

The Common Access Token spec supports the **ES256** and **PS256** signing algorithms in addition to HS256. This PR adds both **signing and verification** for those two asymmetric algorithms. HS256 (HMAC-SHA256) behavior is unchanged.

| Algorithm | COSE id | Structure | `sign()` key | `verify()` key |
| --- | --- | --- | --- | --- |
| `HmacSha256` | 5 | COSE_Mac0 | raw symmetric key | raw symmetric key |
| `Es256` | -7 | COSE_Sign1 | PKCS#8 DER private | SPKI DER public |
| `Ps256` | -37 | COSE_Sign1 | PKCS#8 DER private | SPKI DER public |

## Changes

- **`Algorithm` enum** ([header.rs](src/header.rs)): adds `Es256` / `Ps256`, updates `identifier()`/`from_identifier()`.
- **Crypto primitives** ([utils.rs](src/utils.rs)) via RustCrypto (`p256`, `rsa`, `sha2`): ES256 is ECDSA P-256 + SHA-256 (deterministic RFC 6979, 64-byte `r || s` signatures); PS256 is RSASSA-PSS + SHA-256/MGF1 using `OsRng` for the salt (WASM-friendly via `getrandom`).
- **Token wiring** ([token.rs](src/token.rs)): `sign()`/`verify()` dispatch per algorithm; asymmetric algorithms reuse the existing COSE_Sign1 input and emit CWT tag 61 + COSE_Sign1 tag 18.
- **Key format**: existing `&[u8]` API preserved — PKCS#8 DER private keys for signing, SPKI DER public keys for verifying. New `Error::InvalidKey` ([error.rs](src/error.rs)) for malformed keys.
- **Docs/example**: new [examples/asymmetric_signing.rs](examples/asymmetric_signing.rs), README signing-algorithms section, updated crate docs.
- **Deps**: bump `thiserror` to `2`; bump `rust-version` to `1.88.0` — the real minimum. Because `Cargo.lock` isn't committed, downstream users re-resolve to the latest deps, which pull `minicbor-derive 0.19.4` (uses `let`-chains, stabilized in 1.88) and `zeroize 1.9` (edition 2024, needs 1.85). 1.85/1.86 fail to compile `minicbor-derive`; 1.88 builds and tests cleanly. A pinned MSRV CI job (`dtolnay/rust-toolchain@1.88.0`) guards against silent drift. Note this MSRV reality predates the PR: `main` already depended on `minicbor 2.x`, so its declared `1.60` was already stale — this PR just makes the metadata honest. The crypto stack uses current stable RustCrypto versions (the newer 0.14/0.10 releases are pre-1.0 release candidates).

## Review-driven fixes (commit 079409e)

Grounded in RFC 9052 (COSE structures) and RFC 9053 (algorithm registry):

- **COSE protected-header interop** (addresses the open ES256/PS256 review threads): `from_bytes()` previously discarded the original protected-header bytes and `sign1_input`/`mac0_input` re-encoded the decoded header map. COSE signs the *exact* serialized `protected` bstr (RFC 9052 §4.4 — a re-encode must be byte-identical), so verifying a token whose producer used a different-but-valid CBOR encoding (map ordering, integer width) failed. `Token` now preserves `original_protected_bytes` and reuses them in `sign1_input`/`mac0_input`/`to_bytes` (mirroring the existing payload-bytes handling). Fixes the bug for all algorithms — HMAC had the same latent issue, masked by its Sign1/Mac0 fallback.
- **Single source of truth for algorithm → structure**: new `AlgorithmClass { Mac, Signature }` + `Algorithm::class()` (RFC 9053 §2/§3, RFC 9052 §8.1/§8.2). `to_bytes()` now selects the COSE tag (Mac0=17 / Sign1=18) via an *exhaustive* match instead of a `Some(_)` catch-all, so a future MAC algorithm can no longer be silently mis-tagged as COSE_Sign1 — adding a variant is a compile error until its class is declared. `is_mac()` is reimplemented in terms of `class()`.
- **Dedup**: the byte-identical `sign1_input`/`mac0_input` collapse into one `cose_input(context)` helper; the context string (`"Signature1"`/`"MAC0"`, RFC 9052 §4.4/§6.3) comes from `AlgorithmClass::context()`.

## Lossy-claim verification regression (commit ac8fa56)

Follow-up review caught a regression in the payload-bytes caching introduced above. `get_payload_bytes()` reused the producer's original signed payload bytes only when the *full decoded map* equaled the *current claims projection*. But `Claims` is a lossy projection: a registered claim carried with an unexpected CBOR type is dropped on decode — most notably `aud` (key 3) encoded as a **CBOR array** of audiences, which RFC 8392/7519 permit but `RegisteredClaims.aud: Option<String>` cannot hold. For such an unmutated, validly-signed external token the comparison failed, the payload was re-encoded without the dropped claim, and the genuine signature no longer matched — silently rejecting previously-accepted tokens (reject-valid, not accept-forged).

- **Fix**: detect mutation by comparing the current claims projection against a `baseline_payload_projection` captured at decode time (`Claims::from_map(decoded).to_map()`), instead of against the full original map. Both sides drop the same information, so an unmutated lossy token still matches and keeps its byte-faithful signed bytes, while a real mutation still forces a re-encode. The eager re-encode on the cache-hit path is also dropped.
- **Scope**: correctness only — the array `aud` now verifies and round-trips byte-faithfully, but is still not *readable* via the typed `audience()` accessor. Exposing it would change `RegisteredClaims.aud`'s type (breaking), so it's deferred to a future breaking release.
- **Tests**: `test_aud_as_array_verifies_and_roundtrips`, `test_cti_as_text_verifies_and_roundtrips` (verify + byte-faithful round-trip; fail without the fix), and `test_aud_as_array_mutation_is_reflected` (mutation on a lossy token is still reflected and breaks the original MAC).

## ES256 signature malleability — canonical low-S (commit f54f710)

Follow-up review flagged that ES256 signatures were not normalized to "low-S". ECDSA is malleable: for a valid `(r, s)`, the pair `(r, n - s)` verifies over the same message. RustCrypto's `p256` neither emits low-S on signing nor rejects high-S on verifying, so a third party could derive a second valid signature for an unchanged token, and the signature bytes could not serve as a stable token identity (e.g. a dedup/replay key).

- **Fix** ([utils.rs](src/utils.rs)): `compute_es256` normalizes to low-S before encoding; `verify_es256` rejects high-S (non-canonical) signatures. Per RFC 9053 §2.1 / SEC1 and strict COSE/WebCrypto verifiers, low-S is the canonical form.
- **Tests**: `test_es256_signatures_are_low_s` (every emitted signature is low-S across many messages) and `test_es256_high_s_signature_is_rejected` (constructs the malleable `(r, n - s)` twin of a valid signature and asserts it is rejected).

## PS256 minimum RSA key size + tag-mapping cleanup (commit 5b04be7)

Final review pass. The most material finding was that PS256 imposed no floor on the RSA modulus size: neither the DER decoders nor PSS verification reject undersized keys, so a token signed under a weak key (e.g. 512- or 1024-bit, where a 512-bit modulus is factorable on commodity hardware) would sign and verify normally.

- **Fix** ([utils.rs](src/utils.rs)): `compute_ps256` and `verify_ps256` reject any RSA key whose modulus is below 2048 bits (`MIN_RSA_KEY_BITS`) with `Error::InvalidKey`, on both the signing and verification paths. CHANGELOG security note added. Note this is a behavioral tightening — a consumer relying on sub-2048-bit keys would now be rejected (acceptable for this breaking `0.3.0` release).
- **Tests**: `test_ps256_undersized_key_is_rejected_on_sign` / `..._on_verify` using a real 1024-bit key pair.
- **Cleanup**: the COSE structure-tag mapping (Mac0=17 / Sign1=18) moves onto `AlgorithmClass::cbor_tag()`, beside `context()`, so both wire facts the class determines stay co-located; `to_bytes()` calls it instead of inline literals. Added comments explaining why the `Es256`/`Ps256` arms in `verify`/`sign` are kept separate, and marked the intentionally-duplicated demo keys across the two examples and tests.

## Reject untagged `to_bytes()` with no algorithm (commit af2542d)

Final review comment. `to_bytes()` previously emitted a bare, untagged COSE array when the protected header carried no algorithm — output the crate's own `verify()` then rejects with `InvalidFormat`, so it silently produced an unverifiable token. That state is only reachable by hand-building a `Token` via `Token::new` without an algorithm (the `sign()` path already requires one).

- **Fix** ([token.rs](src/token.rs)): `to_bytes()` now returns `Error::InvalidFormat` in that case, symmetric with `verify()`/`sign()`, so the caller bug surfaces instead of being masked. Test `test_to_bytes_without_algorithm_errors`; CHANGELOG note added.
- The other comment in that review (algorithm-confusion in `verify`) is the already-tracked [FOLLOWUPS.md](FOLLOWUPS.md) *Security* item, consciously deferred — see below.

## Deferred follow-ups

[FOLLOWUPS.md](FOLLOWUPS.md) records review findings intentionally left out of this PR to keep its scope focused: an algorithm-confusion hardening for `verify` (decoder-trust design that deserves its own PR — note it is *not* blocked on a breaking release, an additive `verify_with_algorithm` entrypoint would suffice) and a COSE-tag/`alg` cross-check on decode, the array-valued `aud` readability work (breaking — see above), two hot-path allocation costs in the payload/protected-header caching (the comparison rebuild and the cache-hit clone), and assorted cleanups.

## Tests ([tests.rs](src/tests.rs))

Round-trip sign/verify, COSE_Sign1 tag bytes, wrong-key & cross-algorithm rejection, payload tampering, PSS salt randomization, invalid-key errors, the PS256 undersized-key rejection (sign + verify), identifier round-trips, the `class()`/`context()` mapping, a non-canonical protected-header regression test (mints an external ES256 token with `alg=-7` in 2-byte form and asserts both verify and a byte-faithful `to_bytes()` round-trip; fails without the interop fix), the lossy-claim round-trip regression tests above, and the ES256 low-S canonicalization/rejection tests.

## Verification

- `cargo test` — 55 lib tests + 73 doctests pass
- `cargo clippy --all-targets --all-features -- -D warnings` — clean
- `cargo fmt --check` — clean
- `cargo run --example asymmetric_signing` — ES256 and PS256 both sign and verify

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/token.rs [hunk 1]
```diff
@@ -40,12 +43,23 @@ impl Token {
...
 
-        // For HMAC algorithms, use COSE_Mac0 format with CWT tag
-        if let Some(Algorithm::HmacSha256) = self.header.algorithm() {
-            // Apply CWT tag (61)
-            enc.tag(minicbor::data::Tag::new(61))?;
-            // Apply COSE_Mac0 tag (17)
-            enc.tag(minicbor::data::Tag::new(17))?;
+        // Apply the CWT tag (61) followed by the COSE structure tag. HMAC (MAC)
+        // algorithms use COSE_Mac0 (tag 17); asymmetric signature algorithms use
+        // COSE_Sign1 (tag 18).
```
[the remark this answers is not in the document]
Comment (zshenker) [reply src/token.rs]: Updated the comment to make the untagged case explicit (7b8588464f79f3561127100b0829802cfc1b5768): the `None` branch (no algorithm in the header) intentionally emits the bare COSE array untagged, and `from_bytes` accepts both tagged and untagged input. The comment now documents that.

File: Cargo.toml [hunk 2]
```diff
@@ -11,13 +11,20 @@ repository = "https://github.com/fastly/rust-cat"
 readme = "README.md"
 keywords = ["token", "authentication", "authorization", "cbor", "cose"]
 categories = ["authentication", "cryptography", "web-programming"]
-rust-version = "1.60.0"
+rust-version = "1.88.0"
```
[the remark this answers is not in the document]
Comment (zshenker) [reply Cargo.toml:14]: Good catch — the PR description was stale. The code is correct at `1.88.0`; I've updated the description to match and explain why (uncommitted `Cargo.lock` → downstream re-resolves to `minicbor-derive 0.19.4` which needs let-chains/1.88, and `zeroize 1.9` needs edition 2024/1.85). The pinned `msrv` CI job already builds+tests on `1.88.0`, so all three (Cargo.toml, CI, description) are now consistent.

**Review by jedisct1 — CHANGES_REQUESTED** [review]
Thanks for the follow-up work here. I reviewed the updated PR description, the current diff, the previous discussion and review comments, the affected token/header/utils paths, the new examples and docs, and the current CI state. I also ran the branch locally with `rustup run stable cargo test`, `rustup run stable cargo fmt --check`, `rustup run stable cargo clippy --all-targets -- -D warnings`, and both new examples. Those all passed for me. This is an automated review by Swival (https://swival.dev) using only open-source models, intended to help both the author and the maintainers.

My recommendation is not to merge this quite yet, although I do think the ES256 and PS256 implementation direction now looks broadly reasonable and worth continuing.

The remaining blocker for me is the public compatibility story. This PR changes two public enums that are currently exhaustive: `Algorithm` in `src/header.rs:43-50` gains `Es256` and `Ps256`, and `Error` in `src/error.rs:7-31` gains `InvalidKey`. Downstream users can legally have exhaustive matches over both of those enums today, and those users will fail to compile once these variants are added. The PR also intentionally raises `rust-version` from 1.60 to 1.88 in `Cargo.toml:14`, which is another compatibility break even though the description now explains why that is necessary.

That may all be acceptable for this feature, but it should not be accidental. `Cargo.toml:3` still says `version = "0.2.7"`, so if this is released as another 0.2.x patch, users with normal compatible-version requirements on 0.2 can pick it up and get source breaks. Before merge, I think the PR should either make this a clearly intentional 0.3.0-style breaking release, or the maintainers should separate the code change from release versioning in a way that leaves an explicit note that this cannot be shipped as a 0.2.x patch. Another option would be to first make these public enums non-exhaustive in a deliberate breaking release, but as the code stands, the compatibility impact needs to be handled directly.

There is also one smaller unresolved test-quality issue from the current review thread. `test_es256_tampered_payload_fails` in `src/tests.rs:1595-1613` still accepts `Token::from_bytes` returning `Err` and then performs no assertion. That means the test can pass without proving that a structurally valid token with a tampered payload is rejected by ES256 verification. This is not the main reason I am requesting changes, but for a crypto behavior change it would be much better to mutate a known payload byte, assert that decoding succeeds, and then assert that verification fails.

I would keep the PR open rather than close it. The core wiring in `src/token.rs` now follows the existing COSE input pattern, the new helpers in `src/utils.rs` use the expected key formats and signature encodings, CI is green including the new MSRV job, and the earlier demo-key and stale-description feedback has been addressed. Once the compatibility/release handling is made explicit, and ideally the tamper test is tightened, I would be comfortable re-reviewing this for merge.

File: src/tests.rs [hunk 3]
```diff
@@ -1382,3 +1382,299 @@ fn test_signed_integer_in_nested_structures() {
...
+        .expect("encoded token should be at least 70 bytes");
+    token_bytes[idx] ^= 0xFF;
+
+    // Decoding may still succeed structurally, but verification must fail.
+    if let Ok(decoded) = Token::from_bytes(&token_bytes) {
+        assert!(
+            decoded.verify(&public_key).is_err(),
+            "Verification should fail for a tampered ES256 token"
+        );
+    }
```
[the remark this answers is not in the document]
Comment (zshenker) [reply src/tests.rs]: Fixed in 158bb17. The test now mutates a byte *inside* the custom string claim contents (XOR with 0x01 to stay in ASCII so the text stays valid UTF-8 and the CBOR structure is intact). It then asserts `from_bytes` succeeds before asserting `verify` fails, so the test can no longer pass without exercising signature rejection.

**Review by jedisct1 — APPROVED** [review]
Thanks for the follow-up changes here. I reviewed the current PR description, the latest diff at 158bb17, the earlier discussion and review comments, and the touched implementation paths in `src/header.rs`, `src/token.rs`, `src/utils.rs`, `src/error.rs`, the new tests, examples, README, and CI workflow. This is an automated review by Swival (https://swival.dev) using only open-source models, intended to help both the pull request author and the maintainers.

My recommendation is to merge this PR now.

The previous compatibility blocker looks resolved in the current revision. `Cargo.toml` now makes the breaking release explicit with `version = "0.3.0"` and `rust-version = "1.88.0"`, and the new MSRV job in `.github/workflows/ci.yml` checks that compiler version directly. The public enum changes in `src/header.rs` and `src/error.rs` are still source-breaking for users with exhaustive matches, but that is now treated as part of the 0.3.0 compatibility boundary, and marking both enums `#[non_exhaustive]` prevents the same kind of break from recurring for future algorithm or error additions.

The ES256 and PS256 wiring fits the existing token structure. `TokenBuilder::sign` now signs HMAC tokens with the existing COSE_Mac0 input and asymmetric tokens with COSE_Sign1 input, while `Token::verify` dispatches by the protected `alg` header. The new helpers in `src/utils.rs` use the expected key formats documented in the README and examples: PKCS#8 DER private keys for signing and SPKI DER public keys for verification. The examples also now clearly warn that the embedded private keys are demo material and publicly known.

I also rechecked the test coverage that came up in the earlier comments. The wrong-key ES256 test now uses a valid non-matching P-256 public key, and `test_es256_tampered_payload_fails` now mutates a byte inside a decodable payload and asserts that decoding still succeeds before verification fails. That means the tamper test is actually exercising signature rejection rather than passing because the token became malformed.

For validation, GitHub Actions shows both `build` and `msrv` passing on the latest head. Locally, I ran `rustup run stable cargo fmt --check`, `rustup run stable cargo test`, `rustup run stable cargo clippy --all-targets -- -D warnings`, `rustup run stable cargo run --example asymmetric_signing`, and `rustup run stable cargo run --example sample_es256_ps256_tokens`. All of those passed.

I do not see any remaining correctness, regression, or maintainability issue that should block this. The change is substantial and intentionally breaking, but the current PR makes that explicit and backs the new behavior with focused tests and runnable examples.

File: src/token.rs [hunk 4]
```diff
@@ -149,6 +170,14 @@ impl Token {
                 let mac0_input = self.mac0_input()?;
                 verify_hmac_sha256(key, &mac0_input, &self.signature)
             }
+            Algorithm::Es256 => {
+                let sign1_input = self.sign1_input()?;
+                verify_es256(key, &sign1_input, &self.signature)
+            }
```
[the remark this answers is not in the document]
Comment (zshenker) [reply src/token.rs:214]: Fixed in 079409e. The root cause was exactly this: `from_bytes()` decoded the protected header into a map and discarded the original bytes, then `sign1_input()` re-encoded that map — so a token whose producer used a different-but-valid CBOR encoding (map ordering, integer width) reconstructed a different signed input and failed.

The fix preserves the original bytes: `Token` now carries `original_protected_bytes`, populated by `from_bytes()`, and a new `protected_bytes()` helper returns those exact bytes when present (mirroring the existing `original_payload_bytes` handling). `sign1_input()`, `mac0_input()`, and `to_bytes()` all use it, so verification and re-encoding are byte-faithful to the producer's encoding per RFC 9052 §4.4 ("re-encoded byte string is identical to the decoded byte string").

This applies to all algorithms, not just ES256 — HMAC had the same latent bug, previously masked by its Sign1/Mac0 fallback. New regression test `test_es256_verifies_noncanonical_protected_header` mints an external token encoding `alg=-7` in the non-canonical 2-byte form (`0x38 0x06`) and asserts both verify and a byte-faithful `to_bytes()` round-trip; it fails without the fix.

File: src/token.rs [hunk 5]
```diff
@@ -149,6 +170,14 @@ impl Token {
...
+            Algorithm::Ps256 => {
+                let sign1_input = self.sign1_input()?;
+                verify_ps256(key, &sign1_input, &self.signature)
+            }
```
[the remark this answers is not in the document]
Comment (zshenker) [reply src/token.rs:218]: Fixed in 079409e, same fix as the ES256 thread above. PS256 verification goes through the shared `sign1_input()`, which now uses the preserved original protected-header bytes (`original_protected_bytes` on `Token`, via `protected_bytes()`) rather than re-encoding the decoded header map — so verification reproduces the exact signed `protected` bstr per RFC 9052 §4.4.

**Review by jedisct1 — CHANGES_REQUESTED** [review]
Thanks for the latest updates. I reviewed the new commits on top of the earlier approval, the current PR discussion, the unresolved review comments, the token/header/utils changes, the examples and docs, and the current CI state. This is an automated review by Swival (https://swival.dev) using only open-source models, intended to help both the author and the maintainers.

I would not merge this revision yet. The ES256 and PS256 additions still look broadly worthwhile, and the new protected-header preservation fix is in the right direction, but it leaves the same class of round-trip bug on the payload side.

In `src/token.rs:82`, `Token::to_bytes()` still rebuilds the payload from `self.claims.to_map()` and `encode_map(...)`. For tokens created by this crate that is fine, but for tokens decoded from the wire, `from_bytes()` already preserves `original_payload_bytes`, and verification uses those bytes through `get_payload_bytes()` at `src/token.rs:693`. That means a decoded token can verify correctly, then `to_bytes()` can emit a different payload bstr while keeping the existing signature or MAC. The re-encoded token is then invalid.

This is not just theoretical. I reproduced it with the existing external MAC0 fixture from `test_mac0_token_verification_with_original_bytes` in `src/tests.rs:551`. The decoded token verifies with `testSecret`, but after `let reencoded = token.to_bytes()?`, parsing `reencoded` and verifying it fails with `SignatureVerification`. That is a regression in a public serialization path, and it is especially risky because the new protected-header fix now makes `to_bytes()` appear byte-faithful for one signed field while still changing the other signed field.

The fix should be small and consistent with the rest of the PR: `Token::to_bytes()` should use the same preserved payload bytes that signing and verification use, probably by calling `self.get_payload_bytes()?` instead of rebuilding the claims map there. Please also add a regression test that decodes an externally produced token, verifies it, serializes it again, and verifies the serialized result. The existing fixture in `src/tests.rs:551` is a good candidate because it already demonstrates why preserving original payload bytes matters.

I ran `rustup run stable cargo fmt -- --check`, `rustup run stable cargo test`, `rustup run stable cargo clippy --all-targets -- -D warnings`, and both new examples successfully. I also ran the round-trip repro described above, and that is the reason for this request for changes.

My merge recommendation is request changes for now. I would keep the PR open because the change is valuable and the remaining blocker looks focused, but this should be fixed before merging.

File: src/token.rs [hunk 6]
```diff
@@ -40,19 +51,28 @@ impl Token {
...
         // COSE structure array with 4 items
         enc.array(4)?;
 
-        // 1. Protected header (encoded as CBOR and then as bstr)
-        let protected_bytes = encode_map(&self.header.protected)?;
+        // 1. Protected header (encoded as CBOR and then as bstr).
+        // Reuse the original bytes for a decoded token so the re-encoding is
+        // byte-faithful to the producer's encoding (see `protected_bytes`).
+        let protected_bytes = self.protected_bytes()?;
         enc.bytes(&protected_bytes)?;
```
[the remark this answers is not in the document]
Comment (zshenker) [reply src/token.rs:95]: Fixed in 978ea48. `to_bytes()` now encodes the payload via `get_payload_bytes()` — the same preserved-original-bytes helper that signing and verification already use — so a decoded token's payload bstr is byte-faithful to the producer's encoding rather than re-encoded from the claims map. This closes the asymmetric gap where the protected header was preserved but the payload wasn't.

Added a round-trip regression test (`test_decoded_token_reencodes_without_breaking_signature`) using the external COSE_Mac0 fixture: decode → verify → `to_bytes()` → re-parse → verify. It fails without this fix (the re-encoded token no longer verifies) and passes with it.

File: src/utils.rs [hunk 7]
```diff
@@ -19,6 +19,75 @@ pub fn verify_hmac_sha256(key: &[u8], data: &[u8], signature: &[u8]) -> Result<(
     }
 }
 
+/// Compute an ES256 (ECDSA P-256 + SHA-256) signature over `data`.
+///
+/// `key` must be a PKCS#8 DER-encoded P-256 private key. The returned signature
+/// is the fixed-length 64-byte COSE representation (`r || s`), as required by the
+/// COSE specification (RFC 8152 §8.1).
```
[the remark this answers is not in the document]
Comment (zshenker) [reply src/utils.rs]: Fixed in 424d67d. The doc comment now cites RFC 9053 §2.1, which defines the ECDSA `r || s` fixed-length signature encoding (RFC 8152 was split into RFC 9052 for the COSE structures and RFC 9053 for the algorithms). This keeps the reference consistent with the RFC 9052/9053 citations used elsewhere in the PR.

File: src/tests.rs [hunk 8]
```diff
@@ -1382,3 +1414,392 @@ fn test_signed_integer_in_nested_structures() {
...
+    let token_b = build_signed_token(Algorithm::Ps256, &private_key);
+
+    assert_ne!(
+        token_a.signature, token_b.signature,
+        "PSS signatures should differ due to random salt"
+    );
+
+    token_a.verify(&public_key).expect("token_a should verify");
+    token_b.verify(&public_key).expect("token_b should verify");
+}
```
[the remark this answers is not in the document]
Comment (zshenker) [reply src/tests.rs:1967]: Fixed in d814b78. The test no longer goes through `build_signed_token()` (which called `current_timestamp()` separately per token, so the two tokens could be signed over different iat/nbf/exp claims). Both `token_a` and `token_b` are now built from a shared closure with identical, fixed claims (no `current_timestamp()` calls), so the signed payload is byte-for-byte the same and the PSS salt is the only source of entropy. The `assert_ne!` on the signatures therefore genuinely exercises salt randomization, and both still verify.

File: src/token.rs [hunk 9]
```diff
@@ -655,38 +725,90 @@ impl Token {
...
+    /// protected header, so a mutated header is re-encoded rather than silently
+    /// replaced by the producer's original bytes.
+    fn protected_bytes(&self) -> Result<Vec<u8>, Error> {
+        match self.original_protected_bytes {
+            Some(ref original) if protected_matches(original, &self.header.protected) => {
+                Ok(original.clone())
+            }
+            _ => encode_map(&self.header.protected),
+        }
+    }
```
[the remark this answers is not in the document]
Comment (zshenker) [reply src/token.rs:785]: @copilot The code and all the tests are compiling successfully so have you actually validated this assumption?

File: src/token.rs [hunk 10]
```diff
@@ -32,35 +54,63 @@ impl Token {
...
             // Apply CWT tag (61)
             enc.tag(minicbor::data::Tag::new(61))?;
-            // Apply COSE_Mac0 tag (17)
-            enc.tag(minicbor::data::Tag::new(17))?;
+            let cose_tag = match alg.class() {
+                AlgorithmClass::Mac => 17,       // COSE_Mac0
+                AlgorithmClass::Signature => 18, // COSE_Sign1
+            };
+            enc.tag(minicbor::data::Tag::new(cose_tag))?;
         }
```
[the remark this answers is not in the document]
Comment (zshenker) [reply src/token.rs]: Fixed in af2542d. `to_bytes()` now returns `Error::InvalidFormat("Missing algorithm in protected header")` when the header carries no algorithm, instead of emitting a bare untagged COSE array. This is symmetric with `verify()` and `sign()`, which already require an algorithm, so the caller bug (only reachable via a hand-built `Token::new`) now surfaces immediately rather than producing an unverifiable token. Added `test_to_bytes_without_algorithm_errors` and a CHANGELOG note.

File: src/token.rs [hunk 11]
```diff
@@ -149,6 +211,14 @@ impl Token {
                 let mac0_input = self.mac0_input()?;
                 verify_hmac_sha256(key, &mac0_input, &self.signature)
             }
```
[the remark this answers is not in the document]
Comment (zshenker) [reply src/token.rs:205]: Valid, and tracked. This is the algorithm-confusion item already recorded in [FOLLOWUPS.md](FOLLOWUPS.md) under *Security*, and it was consciously deferred for this PR (the maintainer reviewed and approved with it deferred).

To be clear about the reasoning — and I've updated the FOLLOWUPS note to say so explicitly: this is **not** blocked on a future breaking release. 0.3.0 is already breaking, and a minimal mitigation (an additive `verify_with_algorithm(key, expected)` entrypoint that rejects MAC-vs-asymmetric mismatches) would not break existing callers. What's deferred is the *design decision* about where the expectation should live — a dedicated entrypoint vs. threading it through `VerificationOptions`, plus the default-strictness question — which deserves its own focused PR rather than riding along here. The exposure and fix direction are documented so it can be picked up independently and soon.

**Review by jedisct1 — CHANGES_REQUESTED** [review]
Thanks for the continued work here. I reviewed the current PR #5 head at `af2542d`, the changed implementation and tests around `src/token.rs`, `src/header.rs`, `src/utils.rs`, the README and examples, the latest discussion including the new Copilot comments, the CI state, and the stacked PR #6 context. This is an automated review by Swival (https://swival.dev) using only open-source models, intended to help both the pull request author and the maintainers.

My merge recommendation is request changes. I would not merge PR #5 as it stands.

The earlier serialization and compatibility issues look much better now. The protected-header and payload byte preservation paths in `src/token.rs` are covered by regression tests, the lossy-claim round-trip tests address the availability regression we discussed, ES256 low-S handling and the PS256 RSA key-size floor are both sensible, and the new `to_bytes()` error for a missing algorithm fixes the untagged-token corner case. I also verified the current head locally with `rustup run stable cargo fmt -- --check`, `rustup run stable cargo test`, `rustup run stable cargo clippy --all-targets -- -D warnings`, and both new examples. GitHub Actions is green for `build` and `msrv`.

The remaining blocker is a security issue in the public verification API. `Token::verify(&[u8])` at `src/token.rs:201` still chooses the cryptographic primitive from the token's own protected `alg` header, then applies the caller's one byte slice as whatever key type that header names. In the HMAC arm at `src/token.rs:207-218`, those bytes are treated as a shared MAC secret. In the ES256 and PS256 arms at `src/token.rs:225-231`, the same API expects public key bytes.

That becomes unsafe as soon as this PR adds asymmetric verification. A relying party following the new docs can accept ES256 or PS256 tokens by calling `decoded.verify(public_key)`, as shown in `README.md:164-191`, `examples/asymmetric_signing.rs:92-96`, and `examples/sample_es256_ps256_tokens.rs:94-96`. An attacker can instead create a fresh token with `alg = HmacSha256` and compute the HMAC using those public key bytes as the HMAC key. Since the public key is public, the attacker has everything needed to produce a MAC that this `verify()` implementation will accept.

The fact that `alg` is in the protected header does not prevent this attack, because the attacker is not modifying an existing ES256 token. They are minting a new HMAC token whose protected header and MAC are internally consistent, while the application believed it was verifying an asymmetric token with a public key. That is a real authentication bypass for consumers who adopt the new API exactly as the PR documents it.

`FOLLOWUPS.md:8-31` acknowledges this and defers it, but I do not think this is safe to leave as a follow-up after merging PR #5. This PR is the change that introduces and documents the asymmetric `verify(public_key)` path, and it is already framed as the breaking `0.3.0` release. Before this lands, the branch should include an API that binds the expected algorithm or key type to verification, and the README, examples, and tests should use that safe path.

I also checked PR #6, which is directly relevant here. The late Copilot comment on the algorithm-confusion issue is valid in substance, but it seems to have missed that PR #6 is already open against this branch and implements the typed `VerifyingKey` / `verify_with_key` direction. That context matters. I would treat PR #6, or an equivalent fix, as something to fold into this branch before PR #5 merges, not as an optional later hardening item.

The two newest Copilot comments on `to_signed_payload_bytes()` wording and the extra `claims.to_map()` allocation are much smaller. The documentation wording at `src/token.rs:62-71` could be clarified because byte-preserving decoded tokens are not canonical across producers, and the allocation note is a reasonable performance cleanup. Neither of those changes my merge recommendation. The security issue above does.

I would keep this PR open rather than close it, because the ES256 and PS256 implementation is substantial and now has a lot of useful test coverage. But PR #5 should not merge until the algorithm-confusion fix from PR #6, or an equivalent expected-algorithm verification design, is included in the code and docs that will land together.
